Fixed #28889 -- Prevented double submission of admin forms.

mgaligniana opened the pull request:
This is the "You already submitted (alert)" approach for [ticket](https://code.djangoproject.com/ticket/28889)

To run tests please do: `./runtests.py admin_views.test_prevent_double_submission --selenium=chrome,firefox`

Thanks!

kezabelle commented on django/contrib/admin/static/admin/js/change_form.js:
```diff
@@ -2,8 +2,19 @@
+        form.addEventListener('submit', function(event) {
+            event.preventDefault();
+            submitButtons.forEach(button => {
+                button.setAttribute('disabled', true);
```
I don't know enough of the _details_ on this kind of thing (perhaps someone in @django/accessibility can enlighten me), but is this a scenario where we'd want to also set `aria-disabled=true`?

knyghty replied:
As I understand it... `aria-disabled` doesn't affect the application logic at all, so you can still submit the form if it's present. The screen reader will just announce that it's "dimmed" or similar, but it's still interactive. On the other hand, `disabled` will make the screen reader skip the button entirely - the user wouldn't know the button exists - so there is no point using them both together.

I played around a bit with a toy example locally and it seems ok to me like this. If the request takes a really long time it could be confusing I suppose - but I think that's the same for any long form submission. Probably something can be done to alleviate this but I'm not sure it's worth spending the time on in this PR.

carltongibson commented:
buildbot, test on selenium.

mgaligniana commented:
Thank you @carltongibson for the answer!

> What's your thought?

While I was working on the patch I thought that using AJAX would be great for this solution, but it's totally a new feature so I discarded it

> Do you think there's a suitably more sophisticated (yet still simple) option available? What might that look like?

I just added one more test for the "submit and go back" and it works in Firefox and Chrome, but not in Safari (as you showed in the video) and the solution that comes to my mind now is adding a "submitting" value in the localStorage (something similar like the `navSidebarIsOpen` logic)

> Perhaps you take it this is the right approach, despite the issue? Maybe we should decide to wontfix the issue?

I think we have a good approach, but I don't know why in Safari is not working and it scares me because: _Then the option would be to make the JS more complicated, and off down the rabbit-hole we go._

Also, in my last commit you can see how Firefox handles very well the issue. Maybe we should keep the test where we show that Safari and Chrome create duplicates and ... who knows, in the future maybe, Chrome and Safari will handle/fix it too?

thibaudcolas commented:
I suspect the behavior shown by @carltongibson is due to Safari’s [bfcache](https://web.dev/bfcache/). If so, we could use an appropriate event in the page lifecycle to re-enable the buttons (say on `pageshow` or an event fired before the page gets cached). I’m surprised this wouldn’t also happen with Firefox and Chrome, although perhaps their bfcache support is much more recent.

If we wanted to support the "form submission fails" and "navigation is cancelled" scenario, another option would be to prevent the double submit in a way that stays under the control of end users:

1. Rather than disabling the button, have every submit after the first trigger an additional confirmation dialog (could be as simple as a `window.confirm`).
2. Or disable the button with a custom UI component that can be dismissed by end users.

thibaudcolas commented on django/contrib/admin/static/admin/js/change_form.js:
```diff
@@ -2,8 +2,19 @@
 {
     const inputTags = ['BUTTON', 'INPUT', 'SELECT', 'TEXTAREA'];
     const modelName = document.getElementById('django-admin-form-add-constants').dataset.modelName;
+    const submitButtons = document.querySelectorAll('input[type=submit]');
```
Should this be future-proofed by also targetting `button` elements?

mgaligniana replied:
Hi @thibaudcolas! I didn't understand the question, sorry. I just considered the `submit_line.html` file, which includes the 3 `input` `type=submit` "buttons" or there can be others?

thibaudcolas replied:
No worries! I think I’m probably overthinking this to be honest. I’m surprised we’re using `<input type="submit">` for this rather than `<button type="submit"></button>`, but  we’ll worry about this when we get there.

carltongibson commented:
> Rather than disabling the button, have every submit after the first trigger an additional confirmation dialog (could be as simple as a window.confirm).

This may be sufficient? 🤔

mgaligniana commented:
> > Rather than disabling the button, have every submit after the first trigger an additional confirmation dialog (could be as simple as a window.confirm).
> 
> This may be sufficient? thinking

I think it's a good idea, but isn't this a radical change from what we are used to? I mean, adding a second click could be become tedious?

carltongibson commented:
I think the idea is it only comes in **after** the initial submission, so the form is not broken, but you get a confirmation "You already submitted. Did you mean to submit again?"

mgaligniana commented:
Thank you @carltongibson and @thibaudcolas!

I just added the `alert` approach and updated the tests, including when "goes back"

I don't have a Mac to check the selenium tests but I borrowed an iPad and I noticed that the alert appears in a different moment! Here is the video:

https://user-images.githubusercontent.com/16822952/146965794-8e6aa539-466e-40b3-9b32-21776c574391.MP4

What do you think?

thibaudcolas commented on django/contrib/admin/static/admin/js/change_form.js:
```diff
@@ -2,8 +2,21 @@
+        form.addEventListener('submit', function(event) {
+            event.preventDefault();
+            if (submitted) {
+                const answer = window.confirm('You already submitted. Did you mean to submit again?');
```
This will need to be translated – I believe you’ll need `/*global gettext*/` at the top of the file, and here:

```suggestion
                const answer = window.confirm(gettext('You already submitted. Did you mean to submit again?'));
```

I have no experience with this myself, just trying to match a similar implementation at https://github.com/django/django/blob/main/django/contrib/admin/static/admin/js/actions.js#L161.

mgaligniana replied:
Thank you!!

thibaudcolas reviewed (commented):
Your iPad recording of the "go back" scenario matches my testing and that’s the behavior I was expecting. My testing in macOS Monterey 12.0.1:

- In Safari TP 136 (Safari 15.4), the page gets cached, so on "go back" after submit the `window.confirm` quicks in.
- In Firefox 95, the page doesn’t get cached but the form does get pre-filled with the values I just submitted
- In Chrome Canary 99.0.4765.0, the page doesn’t get cached, and similarly the form will be pre-filled

Running Chrome’s back/forward cache tester in the DevTools, it says it’s bypassed caching because the page has an `unload` event listener and also the `cache-control: no-store` header.

So all good! We’ll just want to remove this "go back" test in Selenium if there is no way to test on Safari there.

mgaligniana commented:
Translate done!

Waiting to decide what we should do with the "go back" test.

mgaligniana commented on tests/admin_views/test_prevent_double_submission.py:
```diff
@@ -0,0 +1,91 @@
+        alert.accept()
+        self.wait_page_ready()
+
+        OBJECTS_CREATED = 1
```
I also wonder if this conditional is correct

claudep commented on django/contrib/admin/static/admin/js/change_form.js:
```diff
@@ -1,9 +1,23 @@
     if (modelName) {
         const form = document.getElementById(modelName + '_form');
+
+        form.addEventListener('submit', function(event) {
```
May we use an arrow function here?

mgaligniana replied:
Great! I just updated in the second PR with the [disable buttons approach](https://github.com/django/django/pull/15229)

mgaligniana replied:
I agree with [thibaudcolas comment](https://github.com/django/django/pull/15217#issuecomment-1018935149) and think we should continue with this approach so I applied the change in this PR too!

Thank you!

claudep commented on django/contrib/admin/static/admin/js/change_form.js:
```diff
@@ -1,9 +1,23 @@
+        form.addEventListener('submit', function(event) {
+            event.preventDefault();
+            if (submitted) {
+                const answer = window.confirm(gettext('You already submitted. Did you mean to submit again?'));
```
I'm not a native speaker, but is `You already submitted` explicit enough? Or `You already submitted this form`, `You already saved this form content`, ... There are better writer than me!

knyghty replied:
I'd probably suggest something along the lines of `You have already submitted this form. Are you sure you want to submit it again?`

mgaligniana replied:
PR updated and change applied! Thank you!

claudep commented on django/contrib/admin/static/admin/js/change_form.js:
```diff
@@ -1,9 +1,23 @@
+            event.preventDefault();
+            if (submitted) {
+                const answer = window.confirm(gettext('You already submitted. Did you mean to submit again?'));
+                if (!answer) {return;}
```
I don't think the braces are really useful. Are they enforced by a linter?

mgaligniana replied:
Correct!

adamchainz replied:
We have eslint’s [curly rule](https://eslint.org/docs/rules/curly) active:

https://github.com/django/django/blob/78f062f63e7dea09c219fd1310d43950817f4c78/.eslintrc#L5

It prevents accidents from missing braces

adamchainz reviewed (commented):
Rather than the confirm dialog approach, can we not re-enable the button on the [`pageshow` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/pageshow_event)? This is triggered when the page is navigated to via "back".

mgaligniana commented:
> Rather than the confirm dialog approach, can we not re-enable the button on the [`pageshow` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/pageshow_event)? This is triggered when the page is navigated to via "back".

Sounds good! Can I create a second branch based on the first approach (disable buttons) to see how it works and can compare?

adamchainz commented:
> Can I create a second branch based on the first approach (disable buttons) to see how it works and can compare?

Sure that's a good idea.

thibaudcolas commented:
I guess it’s a question of whether the feedback shared here https://github.com/django/django/pull/9425#pullrequestreview-81188407 is indeed something worth catering for, or not.	

> I'm always a little unsure of these things. What if the form submission fails or the navigation is cancelled? This will leave the form unusable which can be immensely annoying!

To my knowledge there is no way to listen for cancellation or network failure of requests triggered by HTML forms.

adamchainz commented:
> To my knowledge there is no way to listen for cancellation or network failure of requests triggered by HTML forms.

I just poked around and yeah this is odd.

Perhaps we could instead use a timeout to reenable the form?

mgaligniana commented:
> > Can I create a second branch based on the first approach (disable buttons) to see how it works and can compare?
> 
> Sure that's a good idea.

I just created the [buttons disabled approach](https://github.com/django/django/pull/15229) in a separate PR. I tested in an iPad and it works as expected.

I hope I didn't make a mess with these two PR! But now we have the two alternatives (until now) well divided. Let me know any update/change needed!

Thank you all for your help and opinions!

thibaudcolas commented:
My preference is still for the approach in this PR, as disabled buttons are [known to confuse users](https://axesslab.com/disabled-buttons-suck/) (I’d summarise this as “Clicking doesn’t work, I don’t know, I don’t know what to do about it“). The type of scenarios we’re discussing here don’t feel very common, but nonetheless I find the alternatives more compelling even if more complex.

carltongibson commented:
Thanks @mgaligniana again for your work here, and everyone for the discussion so far!

Can I just check the consensus, and we'll see if we can get this 
moved forward?

These JavaScript issues are always tricky to form an opinion on. 
We're stepping out of our confort zones. Whilst we can do that, I think 
there's a natural (hence vague) complexity line that we don't want to go beyond. 

From a maintenance perspective I also have a certain risk-aversion. I can't 
help ask, _Is this likely to end in a release blocker, bug-in-a-new-feature say?_

On both those counts I find myself leaning towards the `alert()` based
option here, rather than the _disable the buttons_ option over in
#15229. That too looks simple enough as it is, but I'm concerned
they'll be issues that we've not considered that we'd then be on the
hook to handle. I **suspect** that a fully robust version would go
beyond our small JS complexity budget.

Can I check, are we happy to proceed on that basis? @thibaudcolas is I
think in favour? @adamchainz, @claudep, @ngnpope, @kezabelle? You don't
all have to answer, but if you feel strongly do say! 🙂
Thanks! 

If we are going to use the approach here: 

* Isn't `alert()` deprecated? The new [dialog element has broad support](https://caniuse.com/dialog). Should we use that? 
* If we do use `dialog`, we have a bit more freedom over what's displayed, so can we say what the two options entail? (If you resubmit the form, the previous data will be overridden, or you can refresh/navigate away instead... — obviously, exact phrasing there.) — I don't suppose the alert will be shown that often but I imagine the current message leaving room for head-scratching. 

See also [Introducing the Dialog Element](https://webkit.org/blog/12209/introducing-the-dialog-element/) on the Webkit blog.

claudep commented:
Just a quick note about dialog, browser support is too recent in my opinion to use it now.

carltongibson commented:
@claudep [All our supported browsers are green](https://caniuse.com/dialog) ? 🤔

claudep commented:
But you also have to take into account LTS versions (e.g. Firefox ESR). Safari support is also very recent. In a ideal world, all Internet users would have the latest browser version, but there are many things that prevent that, as you can see when you read the "Usage" part on caniuse.com. I would'nt like to be part of the planned obsolescence pushers :-P

adamchainz commented:
I'm in favour of this approach over the disable-buttons approach - the article @thibaudcolas pointed to made me rethink whether it's a good idea to (ever) disable buttons.

I'd be happy using `<dialog>` using [the Google polyfill](https://github.com/GoogleChrome/dialog-polyfill). `alert()` isn't going anywhere but [it has a bunch of caveats](https://developers.google.com/web/updates/2017/03/dialogs-policy).

kezabelle commented:
I think I'm probably also in favour of not _disabling_ the buttons. It's fundamentally pretty intractable unless the whole application is offline-first. Network latency, flakey networks, middleboxes, server-downtime etc are all out of the user's control but each can necessitate a retry. Mostly when you disable buttons it's with the expectation that something will happen within N period, and you'd re-enable the buttons thereafter (or after X/Y/Z anticipated errors, more than half of which are probably _networky_ or bf-cache related). There's not really a once-size-fits-all solution to that in the admin because downstream developers could put _anything_ behind the POST request, even if only in pathological cases.

Ultimately the purpose of this is to provide _feedback_ that you're (as a user) probably doing the wrong thing, and I think disabling the buttons maybe takes that understanding away, where using some form of modal blocker that allows for an escape-hatch gives you both feedback _and_ the flexibility to summarily ignore it (to your own detriment, perhaps).

I'd say I'm probably also in favour of using `<dialog>` over alert/confirm _but_ I think doing so without a polyfill is the more progressive-enhancement (and lightweight?) route, if it's achievable. AFAIK you can detect whether `<dialog>` and it's JS API are available _via_ JS, so if the necessary components are available in `window/globalThis` defer to using the `<dialog>` and if they're not, fall back to `alert/confirm` - though I expect you'd have to conditionally hide the `<dialog>` output if it's not supported...

The only caveat for `<dialog>` I can think of is that _I personally_ don't know how it renders on mobile devices. I'd _presume_ it's well-specced by the browsers and each implement it sensibly, but I don't think I've seen it in the wild.

thibaudcolas commented:
The main reason I proposed window.confirm is simplicity – in LOC terms it’s very similar to disabling the buttons since it’s a blocking call. A custom dialog would be nice too, but it feels like one or two orders of magnitude more JS/HTML/CSS. I’ve never worked with `<dialog>` myself, it seems promising, but from what I’ve read ([Having an open dialog](https://www.scottohara.me/blog/2019/03/05/open-dialog.html)) it doesn’t seem to be ready for prime time (in particular the polyfill has known accessibility issues).

On browser support, personally for a project like Django I’d recommend Safari 12 or 13 as the oldest browser versions to try to support at this point in time (see [support comparison tables](https://caniuse.com/?compare=safari+12.1,safari+13.1&compareCats=all) for the two). That’s "only" 3 years ago, so it’s a reasonable burden considering lots of people would be stuck on those due to Safari releases being partly tied to the OS.

---

Personally I’d recommend going ahead with the PR as it is – it did the job pretty well when I tested it a few commits ago. If we’re still keen on a more elegant dialog – that could be a nice ticket in its own right (would require a bit of research on which library to use, or how to combine `<dialog>` and the polyfill  and the documented fixes for the polyfill’s issue).

carltongibson reviewed (approved):
OK, super. Thanks all. 

Let's go with it as is (wording was adjusted) — Follow ups are always possible: this ticket is four years old, so sometime around the end of the decade would be about right. 😜

I'll rebase and whatnot.

carltongibson commented:
buildbot, test on selenium.

carltongibson commented:
buildbot, test on selenium.
